python/tank/context.py

```diff
@@ -456,16 +460,36 @@ def as_template_fields(self, template):
                                 "does not have any associated folders created on disk yet and "
                                 "therefore no template data can be extracted. Please run the folder "
                                 "creation for %s and try again!" % (self, self.shotgun_url))
-            
+
             # first look at which ENTITY paths are associated with this context object
             # and use these to extract the right fields for this template
             fields = self._fields_from_entity_paths(template)
-            
-            # Determine field values by walking down the template tree
-            fields.update(self._fields_from_template_tree(template, fields, entities))
+
+            # filter the list of fields to just those that don't have a 'None' value.
+            # Note: A 'None' value for a field indicates an ambiguity and was set in the 
+            # _fields_from_entity_paths method (!)
+            non_none_fields = dict([(key, value) for key, value in fields.iteritems() if value is not None])
+
+            # Determine additional field values by walking down the template tree
+            fields.update(self._fields_from_template_tree(template, non_none_fields, entities))
 
         # get values for shotgun query keys in template
         fields.update(self._fields_from_shotgun(template, entities))
+
+        if error_on_missing_fields:
+            # check that all context template fields were found and if not then raise a TankError
+            missing_fields = []
+            for key_name in template.keys.keys():
+                if key_name in entities and key_name not in fields:
+                    # we have a template key that should have been found but wasn't!
+                    missing_fields.append(key_name)
+
+            if missing_fields:
+                raise TankError("Cannot resolve template fields for context '%s' - the following "
+                                "keys could not be resolved: '%s'.  Please run the folder creation "
```

Comment:
Is the "please run the folder creation again" thing valid? what if the template requires more fields than the context currently contains?